### PR TITLE
Flickr30k

### PR DESCRIPTION
**Pull request by jacob-morrison**
Adding a dataset reader, model, and training config for flickr30k image retrieval

### CODE REVIEW

File: allennlp_models/vision/models/vilbert_image_retrieval.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                rel_scores, indices = scores.topk(self.k, dim=-1)
+
+                # Shape: (batch_size)
+                pre_logits = torch.sum((indices == label.reshape(-1, 1)), dim=-1).float()
+
+                # Shape: (batch_size, 2)
+                # 1st column == 1 if we found the image in the top k, 2nd column == 1 if we didn't
+                logits = torch.stack((pre_logits, (pre_logits == 0).float()), dim=1)
+
+                outputs = {"logits": logits}
```
Comment (dirkgr) [inline allennlp_models/vision/models/vilbert_image_retrieval.py:126]: That's not what logits are. In this code, the "scores" would be the closest thing to logits? I don't know if that term makes sense here at all.

Comment (dirkgr) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:126]: Did you call it that for compatibility with some other code?

Comment (dirkgr) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:126]: Oh, I see. They are logits in the training code path, and it makes sense there.

Comment (dirkgr) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:126]: The `CategoricalAccuracy` metric already supports top-k. Can you just use that instead of rolling your own?

File: allennlp_models/vision/models/vilbert_image_retrieval.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            label_namespace,
+            is_multilabel=False,
+            ignore_text=ignore_text,
+            ignore_image=ignore_image,
+        )
+        self.classifier = torch.nn.Linear(pooled_output_dim, 1)
+
+        self.accuracy = CategoricalAccuracy()
+        self.loss = CrossEntropyLoss()
+        self.fbeta = FBetaMeasure(beta=1.0, average="macro")
```
Comment (dirkgr) [inline allennlp_models/vision/models/vilbert_image_retrieval.py:74]: Do you use this anywhere?

File: allennlp_models/vision/models/vilbert_image_retrieval.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        self, output_dict: Dict[str, torch.Tensor]
+    ) -> Dict[str, torch.Tensor]:
+        batch_tokens = []
+        for batch_index, batch in enumerate(output_dict["probs"]):
+            tokens = {}
+            for i, prob in enumerate(batch):
+                tokens[self.vocab.get_token_from_index(i, self.label_namespace)] = float(prob)
+            batch_tokens.append(tokens)
+        output_dict["tokens"] = batch_tokens
+        return output_dict
```
Comment (dirkgr) [inline allennlp_models/vision/models/vilbert_image_retrieval.py:161]: This is copy&paste gore, isn't it? That looks like the function from VQA.

File: allennlp_models/vision/predictors/vilbert_ir.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            raise ValueError(
+                f"Dataset reader is of type f{self._dataset_reader.__class__.__name__}. "
+                f"Expected {Flickr30kReader.__name__}."
+            )
+
+    @overrides
+    def predictions_to_labeled_instances(
+        self, instance: Instance, outputs: Dict[str, numpy.ndarray]
+    ) -> List[Instance]:
+        return [instance]  # TODO
```
Comment (dirkgr) [inline allennlp_models/vision/predictors/vilbert_ir.py:36]: ?

Comment (dirkgr) [reply allennlp_models/vision/predictors/vilbert_ir.py:36]: If you don't have time to finish this, throw `NotImplementedError`.

**Review by dirkgr — CHANGES_REQUESTED** [review]

File: allennlp_models/vision/predictors/vilbert_ir.py (hunk 4 above)
Comment (jacob-morrison) [reply allennlp_models/vision/predictors/vilbert_ir.py:36]: This is copied from the vilbert_vqa predictor, but I think I've fixed it now either way.

File: allennlp_models/vision/models/vilbert_image_retrieval.py (hunk 3 above)
Comment (jacob-morrison) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:161]: Yup, I didn't see anything VQA-specific in it so I thought it looked reasonable, but let me know if I should change it.

File: allennlp_models/vision/models/vilbert_image_retrieval.py (hunk 2 above)
Comment (jacob-morrison) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:74]: No, I thought I deleted it. Thanks!

File: allennlp_models/vision/models/vilbert_image_retrieval.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            return outputs
+
+        else:
+            with torch.no_grad():
+                # Shape: (batch_size, num_images, pooled_output_dim)
+                self.backbone.eval()
+                pooled_output = self.backbone(box_features, box_coordinates, box_mask, caption)[
+                    "pooled_boxes_and_text"
+                ]
+                self.backbone.train()
```
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:111]: Done.

File: allennlp_models/vision/dataset_readers/vision_reader.py [hunk 6]
```diff
@@ -96,11 +96,13 @@ def __init__(
...
         write_to_cache: bool = True,
+        manual_distributed_sharding: bool = True,
+        manual_multiprocess_sharding: bool = True,
     ) -> None:
         super().__init__(
             max_instances=max_instances,
-            manual_distributed_sharding=True,
-            manual_multiprocess_sharding=True,
+            manual_distributed_sharding=manual_distributed_sharding,
+            manual_multiprocess_sharding=manual_multiprocess_sharding,
```
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/vision_reader.py:105]: Yes, I got it to start training.

File: allennlp_models/vision/dataset_readers/vision_reader.py [hunk 7]
```diff
@@ -232,6 +240,36 @@ def _coordinates_cache(self) -> MutableMapping[str, Tensor]:
...
+                self._hard_negative_coordinates_cache_instance = {}
+            else:
+                os.makedirs(self.feature_cache_dir, exist_ok=True)  # type: ignore
+                self._hard_negative_coordinates_cache_instance = TensorCache(
+                    os.path.join(self.feature_cache_dir, "hard_negative_coordinates"),  # type: ignore
+                    read_only=not self.write_to_cache,
+                )
+
+        return self._hard_negative_coordinates_cache_instance
+
```
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/vision_reader.py:272]: Sure! I added it there because the eventual COCO IR reader will need the same stuff, but we can just move it back whenever we write that one.

File: allennlp_models/vision/dataset_readers/flickr30k.py [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+    def text_to_instance(
+        self,
+        caption_dicts: List[Dict[str, Any]],
+        image_index: int,
+        caption_index: int,
+        features_list: List[Tensor] = [],
+        features_list_field: Optional[ListField] = None,
+        coordinates_list: List[TensorField] = [],
+        coordinates_list_field: Optional[ListField] = None,
+        masks_list: Optional[Tensor] = None,
```
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/flickr30k.py:262]: We can (and we'd most likely pick the `List` version), but then we'll remake the ListField for each validation instance. So we'd make a ListField with size 1000 for each of the 5000 validation instances.

File: allennlp_models/vision/dataset_readers/flickr30k.py [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+            masks.append(
+                ArrayField(
+                    hard_negative_features.new_ones(
+                        (hard_negative_features.shape[0],),
+                        dtype=torch.bool,
+                    ),
+                    padding_value=False,
+                    dtype=torch.bool,
+                )
+            )
```
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/flickr30k.py:362]: There's no performance hit, but it wouldn't really simplify the signature. Either we pass in `hard_negative_features` and `hard_negative_coordinates`, or we pass in `averaged_features` and `caption_tensor` to calculate the hard negatives. I think the simplest way to slim down the method signature would be to find a way to pass in only one of each of the List/ListField parameter pairs.

File: allennlp_models/vision/models/vilbert_image_retrieval.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    def __init__(
+        self,
+        vocab: Vocabulary,
+        text_embeddings: TransformerEmbeddings,
+        image_embeddings: ImageFeatureEmbeddings,
+        encoder: BiModalEncoder,
+        pooled_output_dim: int,
+        fusion_method: str = "mul",
+        dropout: float = 0.1,
+        label_namespace: str = "answers",
```
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:51]: I don't believe so, so I think we can remove this parameter.

File: allennlp_models/vision/dataset_readers/flickr30k.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply allennlp_models/vision/dataset_readers/flickr30k.py:262]: Making fields is not expensive. At least, making `ListField` is not expensive. It just keeps a reference to the actual data. So let's just pass in `List`.

File: allennlp_models/vision/dataset_readers/flickr30k.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply allennlp_models/vision/dataset_readers/flickr30k.py:362]: :+1:

File: allennlp_models/vision/models/vilbert_image_retrieval.py (hunk 3 above)
Comment (dirkgr) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:161]: This function assembles `output_dict["tokens"]`, because it thinks the answers it produces are tokens. This won't work in this model at all.

We can have this function do nothing. I don't know exactly how we would represent "the right image" in a `Dict` in a human-readable way. I guess if we had the original filename, that would be neat to add, but it's a bit of work to plumb that all the way through.

Comment (jacob-morrison) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:161]: Oh I see, that makes sense. Thanks for clarifying that. I'll make it just return the output dictionary.

File: allennlp_models/vision/models/vilbert_image_retrieval.py (hunk 1 above)
Comment (jacob-morrison) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:126]: Hm. Will that work if I'm using the accuracy without top-k scoring for the training runs? Or should I create a second accuracy metric based on top-k scoring? And if I should create a second one, can you point me at a good example of how to use it? I'm not sure what its inputs should look like in top-k mode.

File: allennlp_models/vision/dataset_readers/flickr30k.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/flickr30k.py:262]: Sounds good!

File: allennlp_models/vision/dataset_readers/flickr30k.py [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+            coords = [TensorField(coordinates_list[image_index])]
+            masks = [
+                ArrayField(
+                    features_list[image_index].new_ones(
+                        (features_list[image_index].shape[0],), dtype=torch.bool
+                    ),
+                    padding_value=False,
+                    dtype=torch.bool,
+                )
+            ]
```
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/flickr30k.py:303]: I think so too, but I can't remember what the reasoning was? All the other image dataset readers explicitly create the masks, so I assume I should do that here as well.

File: allennlp_models/vision/models/vilbert_image_retrieval.py (hunk 1 above)
Comment (dirkgr) [reply allennlp_models/vision/models/vilbert_image_retrieval.py:126]: Yes, you'd have to have two different metrics, because the metric forces you to specify `k` in the constructor.

File: allennlp_models/vision/dataset_readers/flickr30k.py [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+
+    def __init__(
+        self,
+        image_dir: Union[str, PathLike],
+        *,
+        image_loader: Optional[ImageLoader] = None,
+        image_featurizer: Optional[Lazy[GridEmbedder]] = None,
+        region_detector: Optional[Lazy[RegionDetector]] = None,
+        feature_cache_dir: Optional[Union[str, PathLike]] = None,
+        data_dir: Optional[Union[str, PathLike]] = None,
```
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/flickr30k.py:101]: They're actually slightly different. The `file_path` is a link to a text file that contains the image IDs of all the images in that split, and the `data_dir` contains all of the caption files that are associated with each image. So we read the file at `file_path` to collect a list of image names, and then we read the relevant caption files in `data_dir`. I'll switch `data_dir` to `cached_path`, so we can use a URL.

File: allennlp_models/vision/dataset_readers/flickr30k.py [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+            tokenizer=tokenizer,
+            token_indexers=token_indexers,
+            cuda_device=cuda_device,
+            max_instances=max_instances,
+            image_processing_batch_size=image_processing_batch_size,
+            write_to_cache=write_to_cache,
+            manual_distributed_sharding=False,
+            manual_multiprocess_sharding=False,
+        )
+        self.data_dir = data_dir
```
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/flickr30k.py:127]: Sounds good, I'll test that now.

File: allennlp_models/vision/dataset_readers/flickr30k.py [hunk 14]
```diff
@@ -88,6 +90,12 @@ class Flickr30kReader(VisionReader):
         the captions and metadata for each task instance.
     tokenizer: `Tokenizer`, optional
     token_indexers: `Dict[str, TokenIndexer]`
+    featurize_captions: `bool`, optional
+        If we should featurize captions while calculating hard negatives, or use placeholder features.
+    is_evaluation: `bool`, optional
+        If the reader should return instances for evaluation or training.
+    n: int, optional
+        The number of potential hard negatives to consider.
```
Comment (dirkgr) [inline allennlp_models/vision/dataset_readers/flickr30k.py:98]: Call it `num_hard_negatives` then?

Comment (dirkgr) [reply allennlp_models/vision/dataset_readers/flickr30k.py:98]: Has that always been configurable? I thought we were always using one hard negative?

File: allennlp_models/vision/dataset_readers/flickr30k.py [hunk 15]
```diff
@@ -136,13 +144,55 @@ def __init__(
...
     def _read(self, file_path: str):
         file_path = cached_path(file_path, extract_archive=True)
         files_in_split = set()
+        i = 0
         with open(file_path, "r") as f:
             for line in f:
+                if self.max_instances is not None and i >= self.max_instances:
+                    break
                 files_in_split.add(line.rstrip("\n"))
+                i += 5
```
Comment (dirkgr) [inline allennlp_models/vision/dataset_readers/flickr30k.py]: You can instead write something like this:
```Python
for i, line in enumerate(f):
    ...
```
... and then compare to `I*5`. Then you don't need to maintain that counter variable.

File: allennlp_models/vision/dataset_readers/flickr30k.py [hunk 16]
```diff
@@ -151,64 +201,51 @@ def _read(self, file_path: str):
...
+            processed_images = self._process_image_paths(
+                self.images[filename] for filename in tqdm.tqdm(filenames, desc="Processing images")
+            )
+        except KeyError as e:
+            missing_id = e.args[0]
+            raise KeyError(
+                missing_id,
+                f"We could not find an image with the id {missing_id}. "
+                "Because of the size of the image datasets, we don't download them automatically. "
+                "Please go to https://visualqa.org/download.html, download the datasets you need, "
```
Comment (dirkgr) [inline allennlp_models/vision/dataset_readers/flickr30k.py:215]: This is not the correct URL, is it?

File: tests/vision/dataset_readers/flickr30k_test.py [hunk 17]
```diff
@@ -94,14 +94,14 @@ def test_evaluation_read(self):
...
-            "without",
+            "with",
             "brown",
             "hair",
             "sits",
             "on",
             "edge",
             "of",
-            "cement",
+            "concrete",
```
Comment (dirkgr) [inline tests/vision/dataset_readers/flickr30k_test.py]: What are these changes for?

**Review by dirkgr — CHANGES_REQUESTED** [review]
Only small changes now

File: allennlp_models/vision/dataset_readers/flickr30k.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply allennlp_models/vision/dataset_readers/flickr30k.py]: Let's leave it in the interest of time. Maybe revisit later.

File: tests/vision/dataset_readers/flickr30k_test.py (hunk 17 above)
Comment (jacob-morrison) [reply tests/vision/dataset_readers/flickr30k_test.py]: Some seeds would select a random "wrong" caption, and then accidentally pick the same caption, because the 3 caption files were the same. Then the test would fail because it would compare two captions that should be different, but weren't. I made a few small tweaks to fix it.

File: allennlp_models/vision/dataset_readers/flickr30k.py (hunk 16 above)
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/flickr30k.py:215]: It's not, thanks for catching that

File: allennlp_models/vision/dataset_readers/flickr30k.py (hunk 15 above)
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/flickr30k.py]: Thanks!

File: allennlp_models/vision/dataset_readers/flickr30k.py (hunk 14 above)
Comment (jacob-morrison) [reply allennlp_models/vision/dataset_readers/flickr30k.py:98]: This is a different variable, that I guess doesn't need to be configurable. We first find the `n` (default = `100`) closest images based on the L2 distance between their averaged features, then out of those `n`, we do the actual calculations with the caption vectors. I'm fine with removing it or renaming it, what do you think?

File: tests/vision/dataset_readers/flickr30k_test.py [hunk 18]
```diff
@@ -8,11 +8,13 @@
 from allennlp.modules.vision.grid_embedder import NullGridEmbedder
 from allennlp.modules.vision.region_detector import RandomRegionDetector
 
+import random
 from tests import FIXTURES_ROOT
 
 
 class TestFlickr30kReader(AllenNlpTestCase):
     def setup_method(self):
+        random.seed(12345)
```
Comment (dirkgr) [inline tests/vision/dataset_readers/flickr30k_test.py:17]: What's the deal with this? Why does that make the tests less random?

Comment (jacob-morrison) [reply tests/vision/dataset_readers/flickr30k_test.py:17]: It doesn't, I was testing something but it didn't work. The test is passing locally, so it's hard to debug what's going wrong.
